File: genisites.js
//----------------------------------------------------------------------
// Copyright (c) 2015 Raytheon BBN Technologies
//
// Permission is hereby granted, free of charge, to any person obtaining
// a copy of this software and/or hardware specification (the "Work") to
// deal in the Work without restriction, including without limitation the
// rights to use, copy, modify, merge, publish, distribute, sublicense,
// and/or sell copies of the Work, and to permit persons to whom the Work
// is furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be
// included in all copies or substantial portions of the Work.
//
// THE WORK IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
// HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE WORK OR THE USE OR OTHER DEALINGS
// IN THE WORK.
//----------------------------------------------------------------------

// Strict mode - error undeclared variables
"use strict";

// Establish the geni namespace
var geni;
geni = geni || {};

// Establish the sites sub-namespace
geni.sites = {

    map: undefined,

    inCountries: undefined,

    outCountries: {},

    getNode: function(id) {
        return this.allNodes[id];
    },

    onLoad: function() {
        var url_params = getURLParameters();
        var center_lat = Number(url_params.lat || 38.0);
        var center_lon = Number(url_params.lon || -98.0);
        var zoom = Number(url_params.zoom) || 4;
        geni.sites.setFilters(url_params);
        geni.sites.map = geni.sites.initMap(zoom, center_lat, center_lon);
        $.getJSON('genisites.json', geni.sites.drawSites);
    },

    setFilters: function(url_params) {
        // Add these as properties so we can use the "in" syntax when
        // filtering.
        if (url_params.in) {
            geni.sites.inCountries = {};
            $.each(url_params.in.split(','), function(i, c) {
                geni.sites.inCountries[c] = undefined;
            });
        }
        if (url_params.out) {
            $.each(url_params.out.split(','), function(i, c) {
                geni.sites.outCountries[c] = undefined;
            });
        }
    },

    addSite: function(site) {
        if (site.country in geni.sites.outCountries) {
            // Never show countries on the out list.
            return;
        }
        if (geni.sites.inCountries) {
            // In countries are defined. Only show countries on the list.
            if (! (site.country in geni.sites.inCountries)) {
                // This country is not on the list.
                return;
            }
        }
        var marker = new google.maps.Marker({
            position: new google.maps.LatLng(site.latitude, site.longitude),
            title: site.name,
            icon: "http://maps.google.com/mapfiles/ms/icons/orange-dot.png",
            map: geni.sites.map,
        });
    },

    drawSites: function(data) {
        $.each(data, function(i, site) {
            geni.sites.addSite(site);
        });
    },

    initMap: function(zoom, center_lat, center_lon) {
        var mapOptions = {
            zoom: zoom,
            center: new google.maps.LatLng(center_lat, center_lon),
            mapTypeId: google.maps.MapTypeId.ROADMAP,
            panControl: false,
            zoomControl: true,
            mapTypeControl: false,
            scaleControl: false,
            streetViewControl: false,
            overviewMapControl: false,
            styles: [
                {
                    featureType: "administrative",
                    elementType: "labels",
                    stylers: [
                        { visibility: "off" }
                    ]
                },
                {
                    featureType: "administrative.neighborhood",
                    stylers: [
                        { visibility: "off" }
                    ]
                },
                {
                    featureType: "landscape",
                    stylers: [
                        { visibility: "off" }
                    ]
                },
                {
                    featureType: "poi",
                    stylers: [
                        { visibility: "off" }
                    ]
                },
                {
                    featureType: "road",
                    stylers: [
                        { visibility: "off" }
                    ]
                },
                {
                    featureType: "transit",
                    stylers: [
                        { visibility: "off" }
                    ]
                },
                {
                    featureType: "water",
                    elementType: "labels",
                    stylers: [
                        { visibility: "off" }
                    ]
                }
            ]
        };

        var map = new google.maps.Map(document.getElementById('map-canvas'),
                                      mapOptions);
        return map;
    }
};

// Load google maps and then call the load handler function
google.maps.event.addDomListener(window, 'load', geni.sites.onLoad);

File: genisites.py
#!/usr/bin/env python
#----------------------------------------------------------------------
# Copyright (c) 2015 Raytheon BBN Technologies
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and/or hardware specification (the "Work") to
# deal in the Work without restriction, including without limitation the
# rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Work, and to permit persons to whom the Work
# is furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Work.
#
# THE WORK IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
# OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
# HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
# WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE WORK OR THE USE OR OTHER DEALINGS
# IN THE WORK.
#----------------------------------------------------------------------

#----------------------------------------------------------------------
#
# This program turns domain names into lat/lon points and outputs a
# JSON file suitable for use with genisites.js. The conversion uses a
# geoip2 package and associated database.
#
# INSTALL
# =======
#
# yum install python-pip
# pip install geoip2
# wget http://geolite.maxmind.com/download/geoip/database/GeoLite2-City.mmdb.gz
# gunzip GeoLite2-City.mmdb.gz
#
# SAMPLE
# ======
# 
#   import socket
#   import geoip2.database
#   reader = geoip2.database.Reader('GeoLite2-City.mmdb')
#   ipaddr = socket.gethostbyname('uvm.edu')
#   x = reader.city(ipaddr)
#   x.location.latitude
#   x.location.longitude
#
#
# DOMAIN DATA
# ===========
#
# Domain data is extracted from the Member Authority database in CSV
# format for input into this program. The PostgreSQL command needs to
# be on one line for the \copy command to work:
#
#  -- Select eppn and email for all users.
#  -- This all needs to be on one line to work.
#  \copy (select mma1.value as eppn, mma2.value as email from ma_member_attribute mma1, ma_member_attribute mma2 where mma1.member_id = mma2.member_id and mma1.name = 'eppn' and mma2.name = 'email_address') To '/tmp/eppn-email.csv' with CSV
#
#
# USAGE
# =====
#
#  geocode.py > genisites.json
#
# NOTE: requires file "eppn-email.csv" in the current working directory
#
#----------------------------------------------------------------------

from __future__ import print_function

from collections import defaultdict
import json
import socket
import sys

import geoip2.database

eppn_domains = defaultdict(int)
email_domains = defaultdict(int)

def process_email(x):
    parts = x.split('@')
    if len(parts) != 2:
        print('error: %r' % (x), file=sys.stderr)
    else:
        domain = parts[1]
        rev_domain = domain.split('.')
        rev_domain.reverse()
        if rev_domain[0] in ['edu', 'gr', 'br', 'de', 'it', 'kr']:
            # limit these domains to the top level
            rev_domain = rev_domain[:2]
        if rev_domain[-1] in ['student', 'students']:
            # remove this leading subdomain
            rev_domain = rev_domain[:-1]
        rev_domain.reverse()
        domain = '.'.join(rev_domain)
        email_domains[domain] = email_domains[domain] + 1

def process_eppn(x):
    parts = x.split('@')
    if len(parts) != 2:
        pass
    else:
        domain = parts[1]
        eppn_domains[domain] = eppn_domains[domain] + 1

def process_line(x):
    (eppn, email) = x.split(',')
    eppn = eppn.strip().lower()
    email = email.strip().lower()
    if 'gpolab.bbn.com' in eppn:
        process_email(email)
    else:
        process_eppn(eppn)

with open('eppn-email.csv') as f:
    for line in f:
        process_line(line)

# combine email and eppn domains
for d in sorted(email_domains):
    eppn_domains[d] = eppn_domains[d] + email_domains[d]

#for d in eppn_domains:
#    print d, eppn_domains[d]

def make_site(domain, geodata, count):
    return dict(name=domain,
                count=count,
                latitude=geodata.location.latitude,
                longitude=geodata.location.longitude,
                country=geodata.country.iso_code)

reader = geoip2.database.Reader('GeoLite2-City.mmdb')
domains = []
for d in eppn_domains:
    try:
        ipaddr = socket.gethostbyname(d)
        x = reader.city(ipaddr)
        domains.append(make_site(d, x, eppn_domains[d]))
    except:
        d2 = 'www.' + d
        try:
            ipaddr = socket.gethostbyname(d2)
            x = reader.city(ipaddr)
            domains.append(make_site(d, x, eppn_domains[d]))
        except:
            msg = "DNS FAILURE: %s, %s (count=%d)" % (d, d2, eppn_domains[d])
            print(msg, file=sys.stderr)
        
msg = "Found locations for %d domains" % (len(domains))
print(msg, file=sys.stderr)
print(json.dumps(domains))
